ambry-router/src/main/java/com.github.ambry.router/AdaptiveOperationTracker.java

```diff
@@ -0,0 +1,160 @@
+   *                                otherwise.
+   * @param successTarget The number of successful responses required to succeed the operation.
+   * @param parallelism The maximum number of inflight requests at any point of time.
+   * @param time the {@link Time} instance to use.
+   * @param localColoTracker the {@link Histogram} that tracks intra datacenter latencies for this class of requests.
+   * @param crossColoTracker the {@link Histogram} that tracks inter datacenter latencies for this class of requests.
+   * @param pastDueCounter the {@link Counter} that tracks the number of times a request is past due.
+   * @param quantile the quantile cutoff to use for when evaluating requests against the trackers.
+   */
+  AdaptiveOperationTracker(String datacenterName, PartitionId partitionId, boolean crossColoEnabled, int successTarget,
+      int parallelism, Time time, Histogram localColoTracker, Histogram crossColoTracker, Counter pastDueCounter,
+      double quantile) {
+    super(datacenterName, partitionId, crossColoEnabled, successTarget, parallelism, true);
+    this.datacenterName = datacenterName;
+    this.time = time;
+    this.localColoTracker = localColoTracker;
+    this.crossColoTracker = crossColoTracker;
+    this.pastDueCounter = pastDueCounter;
+    this.quantile = quantile;
+    this.otIterator = new OpTrackerIterator();
+  }
+
+  @Override
+  public void onResponse(ReplicaId replicaId, boolean isSuccessFul) {
+    super.onResponse(replicaId, isSuccessFul);
+    long elapsedTime;
+    if (unexpiredRequestSendTimes.containsKey(replicaId)) {
+      elapsedTime = time.milliseconds() - unexpiredRequestSendTimes.remove(replicaId);
+    } else {
+      elapsedTime = time.milliseconds() - expiredRequestSendTimes.remove(replicaId);
+    }
+    getTracker(replicaId).update(elapsedTime);
+  }
+
+  @Override
+  public Iterator<ReplicaId> getReplicaIterator() {
+    replicaIterator = replicaPool.iterator();
+    return otIterator;
+  }
+
+  /**
+   * Gets the {@link Histogram} that tracks requests to the class of replicas (intra or inter DC) that {@code replicaId}
+   * belongs to.
+   * @param replicaId the {@link ReplicaId} whose request latency is going to be tracked.
+   * @return the {@link Histogram} that tracks requests to the class of replicas (intra or inter DC) that
+   * {@code replicaId} belongs to.
+   */
+  private Histogram getTracker(ReplicaId replicaId) {
+    if (replicaId.getDataNodeId().getDatacenterName().equals(datacenterName)) {
+      return localColoTracker;
+    }
+    return crossColoTracker;
+  }
+
+  /**
+   * An iterator to fetch replicas to send requests to. Respects parallelism and discounts requests that have been
+   * outstanding for a certain amount of time from parallelism.
+   */
+  private class OpTrackerIterator implements Iterator<ReplicaId> {
+
+    @Override
+    public boolean hasNext() {
+      return replicaIterator.hasNext() && (inflightCount < parallelism || isOldestRequestPastDue());
```

Comment:
Discussed this offline - requests are taken out from `unexpiredRequestSendTimes` when they are past due and another replica is acknowledged via `replicaIterator.remove()`.